hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/CaseExpression.java

```diff
@@ -0,0 +1,181 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Implements evaluation of SQL CASE operator.
+ */
+public final class CaseExpression<T> implements Expression<T>, IdentifiedDataSerializable {
+
+    private Expression<Boolean>[] conditions;
+    private Expression<?>[] results;
+    private QueryDataType resultType;
+
+    public CaseExpression() {
+        // No-op.
+    }
+
+    private CaseExpression(Expression<Boolean>[] conditions, Expression<?>[] results, QueryDataType resultType) {
+        this.conditions = conditions;
+        this.results = results;
+        this.resultType = resultType;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static CaseExpression<?> create(Expression<?>[] expressions, QueryDataType resultType) {
+        // Split conditions and expressions.
+        assert expressions != null;
+        assert expressions.length % 2 == 1;
+
+        int conditionCount = expressions.length / 2;
+
+        Expression<Boolean>[] conditions = new Expression[conditionCount];
+        Expression<?>[] results = new Expression[conditionCount + 1];
+
+        int idx = 0;
+
+        for (int i = 0; i < conditionCount; i++) {
+            conditions[i] = (Expression<Boolean>) expressions[idx++];
+            results[i] = expressions[idx++];
+        }
+
+        // Last expression might be null.
+        results[results.length - 1] = expressions.length == idx + 1 ? expressions[idx] : null;
+
+        // Done.
+        return new CaseExpression<>(conditions, results, resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_CASE;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        for (int i = 0; i < conditions.length; i++) {
+            Expression<Boolean> condition = conditions[i];
+
+            Boolean conditionHolds = condition.eval(row, context);
+            if (TernaryLogic.isTrue(conditionHolds)) {
+                return (T) results[i].eval(row, context);
+            }
+        }
+
+        // Return the last result if none conditions were met.
+        Expression<?> lastResult = results[results.length - 1];
+
+        if (lastResult != null) {
+            return (T) lastResult.eval(row, context);
+        } else {
+            return null;
```

Comment:
This edge case is not covered with tests. Can we really have `lastResult == null`?

Reply:
Honestly, no idea, I just copied it from the pre-existing implementation. I will check it.

Reply:
It can't be null, fixed in 6201d4d8804a29f0b5cedbd5dad9d9a541788e2b